Check cluster ID hasn't changed on each request

fasaxc opened the pull request:
This catches the case where the cluster is rebuilt from a backup
or replaced unexpectedly.  Without this change, we'd potentially
watch on an etcd index that was only valid on the old cluster.

* On read, record the cluster ID of the server.  We have to defer loading of
  the resource so we can check the headers before blocking on a
  watch.
* On watch, check the cluster ID hasn't changed.

lavagetto commented:
Thanks a lot, this makes a lot of sense, but I was thinking of going further with this:

- If no cluster ID is given to the client upon instantiation, the first request we do (the one for the machines in the cluster) should set it. 

- Any subsequent request should raise an exception if the cluster id doesn't appear

This would make your code a lot thinner and more beautiful too. I'm fixing some issue with self.machines today so you may want to wait for that - or I can take your patch and modify it.

Very good work anyways!

fasaxc commented:
Hi,

The patch does set the cluster ID if it's not passed in IIRC. I allow it to
be passed in so that it can be transferred from one client to another if
the user needs to recreate the client.

I'm happy for the exception to be raised on any request (i.e. not just
watch requests) but I was worried that unsuspecting users might get tripped
up by the new exception.

-Shaun
On 26 Apr 2015 12:31 am, "Giuseppe Lavagetto" <notifications@github.com>
wrote:

> Thanks a lot, this makes a lot of sense, but I was thinking of going
> further with this:
>
>    -
>
>    If no cluster ID is given to the client upon instantiation, the first
>    request we do (the one for the machines in the cluster) should set it.
>    -
>
>    Any subsequent request should raise an exception if the cluster id
>    doesn't appear
>
> This would make your code a lot thinner and more beautiful too. I'm fixing
> some issue with self.machines today so you may want to wait for that - or I
> can take your patch and modify it.
>
> Very good work anyways!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jplana/python-etcd/pull/92#issuecomment-96336632>.
>

lavagetto commented:
@fasaxc you're right, you set it, but the code right now is a bit complicated tbh. What I as a developer would expect is that the cluster id is an inherent property of my connection to etcd, which can be set when instantiating the object, or explitictly changed with a specifc method call.

This would really simplify your code - I guess I'll try to show you a simplified version whenever I have time.

fasaxc commented:
I was keen to set it on the first request, whatever that is. IIRC the code
doesn't always get the list of machines, depending on the arguments passed
to the constructor.

For my use case I'd be happy if we checked the cluster ID on every request
but I was worried about raising new exceptions from all reads, which breaks
compatibility with existing apps.

If you're not worried about that then there's no need for the flag.

Shaun

On Sun, 26 Apr 2015 23:57 Giuseppe Lavagetto <notifications@github.com>
wrote:

> @fasaxc <https://github.com/fasaxc> you're right, you set it, but the
> code right now is a bit complicated tbh. What I as a developer would expect
> is that the cluster id is an inherent property of my connection to etcd,
> which can be set when instantiating the object, or explitictly changed with
> a specifc method call.
>
> This would really simplify your code - I guess I'll try to show you a
> simplified version whenever I have time.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jplana/python-etcd/pull/92#issuecomment-96523986>.
>

fasaxc commented:
@lavagetto I've removed the condition, it now checks the ID every time, which is much simpler.  I also rebased onto the current master and fixed up the UTs.

I couldn't move the initialization of the cluster ID to the the `machines()` method because it is only called if `allow_reconnect` is `True`.

fasaxc commented:
Sorry to keep changing this but one of my colleagues got tripped up by the new behaviour, he wasn't expecting the EtcdClusterIdChanged error to be persistent and raised on every call.  I think I should adapt it to better match his expectations.  Otherwise, common patterns like this will start failing after this change:

```
client = etcd.Client()
while True:
    try:
        ...
        client.set(...)
    except EtcdException:
        print "Etcd request failed, retrying..."
```

fasaxc commented:
OK, I've updated the PR with that change.

lavagetto commented:
@fasaxc I think the last version of the patch is what I as an user would expect to see.

LGTM and thanks again for the wonderful work.
